src/fanout/emqx_sdv_fanout_dispatcher.erl

```diff
@@ -0,0 +1,278 @@
+        {ok, RefKey, MRef} ->
+            %% do the heavy lifting in the client process
+            %% because mria:dirty_delete/2 may involve RPC
+            %% to core nodes if running in a replica node
+            ok = emqx_sdv_fanout_ids:delete(RefKey),
+            %% find the dispatcher for the VIN and tell it to send the next message
+            Dispatcher = gproc_pool:pick_worker(?DISPATCHER_POOL, VIN),
+            gen_server:cast(Dispatcher, ?ACKED(self(), RefKey, MRef));
+        {error, not_found} ->
+            %% half-transmitted QoS 1 was retried, ignore for now, fix until someone screams
+            %% TODO: scan the ids table for the VIN + RequestId and delete it
+            ?LOG(warning, "acked_unknown_request_id", #{
+                vin => VIN,
+                request_id => RequestId
+            }),
+            ok
+    end.
+
+%% @doc Handle a heartbeat from a vehicle.
+heartbeat(VIN) ->
+    %% check if the VIN has any messages inflight or pending to be sent in the client process
+    %% so to minimize the number or message passing to the dispatcher pool
+    case emqx_sdv_fanout_inflight:exists(self()) of
+        true ->
+            %% there are inflight messages, do not notify the dispatcher
+            %% because the client process will do it when the inflight messages are acknowledged
+            ?LOG(debug, "inflight_messages_detected", #{
+                vin => VIN,
+                trigger => ?TRG_HEARTBEAT
+            }),
+            ok;
+        false ->
+            case emqx_sdv_fanout_ids:next(VIN) of
+                {ok, {_RefKey, _DataID}} ->
+                    %% this is called by the vehicle client process itself,
+                    %% so it's for sure online, directly notify the dispatcher
+                    %% i.e. no need to call notify_dispatcher_if_vehicle_online/3
+                    do_notify_dispatcher(self(), VIN, ?TRG_HEARTBEAT);
+                {error, empty} ->
+                    %% ignore if the VIN has no inflight messages
+                    ?LOG(debug, "no_pending_messages_to_send", #{
+                        vin => VIN,
+                        trigger => ?TRG_HEARTBEAT
+                    }),
+                    ok
+            end
+    end.
+
+insert_ids(VINs, RequestId, DataID) ->
+    Now = now_ts(),
+    lists:foreach(fun(VIN) -> emqx_sdv_fanout_ids:insert(VIN, Now, RequestId, DataID) end, VINs),
+    ?LOG(debug, "insert_new_vins_for_fanout", #{
+        vin => VINs,
+        request_id => RequestId,
+        data_id => DataID
+    }).
+
+notify_dispatchers(VINs) ->
+    _ = proc_lib:spawn(fun() ->
+        set_label(sdv_fanout_notify_dispatchers),
+        do_notify_dispatchers(VINs)
+    end),
+    ok.
+
+-if(?OTP_RELEASE >= 27).
+set_label(Label) ->
+    proc_lib:set_label(Label).
+-else.
+set_label(_Label) ->
+    ok.
+-endif.
+
+do_notify_dispatchers([]) ->
+    ok;
+do_notify_dispatchers([VIN | VINs]) ->
+    case emqx_cm:lookup_channels(VIN) of
```

Comment:
So there is an assumption that clientId = VIN ?

Reply:
yes